### PR TITLE
[DependencyInjection] Optimized initializing container parameters by using immutable arrays

### PR DESCRIPTION
**Pull request by superdav42**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

This PR was inspired by 
https://github.com/composer/composer/pull/5174
which immproved composer's autoloader performance with a large class map by making the class map use constants instead of variable so opcache could optimize the array.

For dumped containers instead of calculating $targetPaths in the contructor we use the constant ``__DIR__`` with the appropriate number of '/..' to create immutable values that can be used in a static array. Because it is constant opcache will not need to execute the array everytime but can copy it from shared memory.

A quick and easy benchmark of a large container of 26000+ lines for parameters shows that this is 5% faster and uses 3% less memory.
```php
$start = microtime(true);
require '../app/cache/dev/appDevDebugProjectContainer.php'; var_dump('new container');

$container = new appDevDebugProjectContainer();

echo ((memory_get_peak_usage() / 1024 )/1024).'mb';
echo 'end time '. ((microtime(true) -$start) * 1000).'ms';
```
Old container:
2.8640975952148mb
end time 1.7151832580566ms
New Container
2.7830810546875mb
end time 1.6288757324219ms

Savings are significantly more when multiple copies of a container are made. Which I have seen happen in poorly written programs.

### CODE REVIEW

**Comment by stof** [comment]
This is a BC break, because it does not return the real path anymore, and some places are relying on getting a real path (they are resolving real paths before setting them in the container).
Your code modifies the value being stored in the container.

so 👎 for this implementation.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -829,6 +829,14 @@ private function startClass($class, $baseClass, $namespace)
         $bagClass = $this->container->isFrozen() ? 'use Symfony\Component\DependencyInjection\ParameterBag\FrozenParameterBag;' : 'use Symfony\Component\DependencyInjection\ParameterBag\\ParameterBag;';
         $namespaceLine = $namespace ? "namespace $namespace;\n" : '';
 
+        // Inject the parameters as a static property so OPcache can optimize it for speed and memory
+        if ($this->container->getParameterBag()->all()) {
+            $parametersArray = $this->exportParameters($this->container->getParameterBag()->all());
+            $parameters = "\n    private static \$parameters = $parametersArray;";
+        } else {
+            $parameters = '';
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: the variable should always be defined (and contain an empty array), as the code still uses it even when the array is empty.

**Comment by nicolas-grekas** [comment]
Same as @stof: this cannot work. The BC break is that people to things like `dirname($someParam)`, but using `..` in such code would break.
Closing as such.
Of course, the performance target is nice, so if you can spot other ways to enhance things, please do.

**Comment by superdav42** [comment]
@stof I see how it could be a BC break. I think I know a way we could avoid it though, might add a but of complexity, I'll try it and make a new PR. I think there's still gains to be made here which would be worth the complexity for such a critical component.

**Pull request closed without merging**